Scalable NAT Instances

laneshetron opened the pull request:
This repo is a fork of a fork, but with this PR I've adapted it to our needs.

- Create `n` NAT instances corresponding to `n` private subnets under an autoscaling group
- Create 2 ENIs for every NAT instance and split each subnet into two routes pointing at these ENIs
- Added a `dry_run` option for standing up all infrastructure without generating routes. This makes it easier to safely transition existing environments off of NAT Gateway.

Successfully load tested this configuration to 7M RPM on 6 subnets. In practice I think the ceiling is even higher than this.

#### Future work:
- It would be simple to bounce detached ENIs from failed instances over to other instances for fault tolerance, but we would need a way of moving them back over when new instances come back up.
- Scale-in/scale-out support
- Even greater scale could be achieved with larger instances and more ENIs. This could either be configurable or dynamic based on autoscaling.

nelg committed:
my fix to get nat instance working

laneshetron committed:
Update init script to grab any available ENI

laneshetron committed:
Fix missing vars parameter

laneshetron committed:
Fix pt 2

laneshetron committed:
Create a NAT instance and ENI for each private subnet

laneshetron committed:
Update ips output

laneshetron committed:
Fix the other output (ugh)

laneshetron committed:
Remove attachment filter

laneshetron committed:
Remove eni templating

laneshetron committed:
Trim double quotes

laneshetron committed:
Attach 2 interfaces to each NAT instance

laneshetron committed:
Fix arithmetic evaluation

laneshetron committed:
Fix arithmetic (for realsies)

laneshetron committed:
Add dry run option for safely cutting over existing infrastructure

laneshetron committed:
Parameterize device-index

laneshetron committed:
herp derp

laneshetron committed:
Balance upper and lower subnets evenly across instances

laneshetron committed:
Fix typo

laneshetron committed:
Set default instance type to t3.medium to support 3 ENIs

masonmahaffey reviewed (commented):
Awesome! How did you load test this?

laneshetron commented:
@masonmahaffey I stood up a new 'load' environment in doppler receive with separate VPCs, etc., and used our usual lambda based load tool to throw traffic at it using legit receive payloads.

mjreed-wbd commented on runonce.sh:
```diff
@@ -1,11 +1,34 @@
+return_code=1
+tag="lower"
+
+aws ec2 modify-instance-attribute --no-source-dest-check \
```
Since you're extracting the same values from the output of `ec2-metadata` multiple times, it would make sense to capture those once into variables and just use those instead:
    
    region=$(/opt/aws/bin/ec2-metadata -z  | sed 's/placement: \(.*\).$/\1/')
    instance_id=$(/opt/aws/bin/ec2-metadata -i | cut -d ' ' -f2)
    ...
    aws ... --region "$region" --instance-id "$instance_id" ...

mjreed-wbd commented on runonce.sh:
```diff
@@ -1,11 +1,34 @@
+# attach 2 interfaces for maximal bandwidth
+for i in {1..2}; do
+    # retry attachment errors in case of contention
+    while [ $return_code -ne 0 ]; do
```
You can shorten this to `while (( return_code )); do`; , since`((`...`))` evaluates a numeric expression and returns true if it's nonzero. If you'd prefer to be explicit for clarity I would still go with `(( return_code != 0 ))`, just because it lets you use the regular not-equals operator instead of `-ne`. I mean, what is this, Fortran?

laneshetron replied:
show me on the doll where Fortran hurt you!

mjreed-wbd commented on runonce.sh:
```diff
@@ -1,11 +1,34 @@
+  --instance-id "$(/opt/aws/bin/ec2-metadata -i | cut -d' ' -f2)"
+
+# attach 2 interfaces for maximal bandwidth
+for i in {1..2}; do
```
Brace expansion is handy for larger ranges but it'd be shorter to just write `for i in 1 2; do` here.

laneshetron replied:
I ended up keeping this syntax in order to template this file with a variable number of interfaces.

mjreed-wbd commented on snat.sh:
```diff
@@ -5,13 +5,16 @@ while ! ip link show dev eth1; do
   sleep 1
 done
 
+sysctl -q -w net.ipv4.conf.all.rp_filter=0
```
This might call for a loop?

    for param in net.ipv4.conf.{all,eth0,eth1,default}.rp_filter; do
        sysctl -q -w "$param=0"
    done

or maybe:

    for interface in all eth0 eth1 default; do
        sysctl -q -w "net.ipv4.conf.$interface.rp_filter=0"
    done

mjreed-wbd commented on main.tf:
```diff
@@ -1,3 +1,10 @@
+locals {
+  instance_count = length(var.private_route_table_ids)
+  # split each subnet into two routes
+  interface_cidrs = flatten([for i in range(local.instance_count) : ["0.0.0.0/1", "128.0.0.0/1"]])
```
would it make any sense ever to have these two CIDR values configurable via vars?

laneshetron replied:
This is something that I've been considering, hence my "future work" item:

> - Even greater scale could be achieved with larger instances and more ENIs. This could either be configurable or dynamic based on autoscaling.

But one of the issues here is that these EC2 instances have a maximum number of interfaces depending on instance type. So for more than just 2 ENIs, for example, we would simultaneously need to scale up to a 'large' or beyond.

laneshetron replied:
I got anxious about not being able to configure this on the fly, so I've added a few commits that implement this. Should also pave the way for making things dynamically scalable in the future.

mjreed-wbd reviewed (approved):
LGTM, just a few comments.

awlawl commented:
One thing to keep in mind that if you are running ec2 instances you must still abide by the CSO AWS and Amazon Linux baselines. It does look like there is a module variable that can be used to include more user_data scripts so that it can install the agents and adjust the OS defaults.

awlawl commented on main.tf:
```diff
@@ -24,18 +31,21 @@ resource "aws_security_group_rule" "ingress_any" {
-  route_table_id         = var.private_route_table_ids[count.index]
-  destination_cidr_block = "0.0.0.0/0"
-  network_interface_id   = aws_network_interface.this.id
+  count = var.dry_run ? 0 : local.interface_count
```
I love the dry run concept here!

laneshetron committed:
Make CIDR routes configurable

laneshetron committed:
runonce.sh cleanup
make snat.sh scalable

laneshetron committed:
Escape terraform interpolation

laneshetron committed:
Bash is fun. we're having fun..

laneshetron commented:
Thank you both for the reviews! I've addressed some of the feedback and added a few commits for making the routes configurable, which was a fairly involved change.

@awlawl Can you point me in the direction of the "CSO AWS and Amazon Linux baselines"?

laneshetron committed:
Export NAT Auto Scaling Group
